addons/account/models/account_move.py

```diff
@@ -2077,11 +2077,15 @@ def _check_duplicate_supplier_reference(self):
         """ Assert the move which is about to be posted isn't a duplicated move from another posted entry"""
         move_to_duplicate_moves = self.filtered(lambda m: m.state == 'posted' and m.is_purchase_document())._fetch_duplicate_reference(matching_states=('posted',))
         if any(duplicate_move for duplicate_move in move_to_duplicate_moves.values()):
-            duplicate_move_ids = list(set(
+            duplicate_move_ids = list({
                 move_id
-                for move_ids in ([current_move_id] + duplicate.ids for current_move_id, duplicate in move_to_duplicate_moves.items() if duplicate)
+                for move_ids in (
+                    [current_move_id.id] + duplicate.ids
+                    for current_move_id, duplicate in move_to_duplicate_moves.items()
+                    if duplicate
+                )
                 for move_id in move_ids
-            ))
+            })
```

Comment:
this is so weird (and it breaks in 17.2 as well, so we need to backport your fix after that)
what about:
```suggestion
            duplicate_move_ids = flatten([k.ids + v.ids for k, v in move_to_duplicate_moves.items() if v])
```
(flatten comes from odoo.tools)
Ping @Levizar, would it be correct or am I missing sth (asking you cause you wrote it :D) ?

Reply:
This kind of works, but it doesn't get rid of duplicate ids. Not sure if that is an issue though? (Also, the flatten tool seems not used in the codebase apart from 2 upgrade scripts, but ...)

Reply:
Oh yeah, just need to add `list(set(...))` to get rid of the duplicates. Regarding flatten, it's indeed not widely used, (I see 3 occurences in master), but it's straightforward. I guess you could avoid using it, but it makes the code much easier to read here.

Reply:
Yes, it's mybad for this bug.
We've changed the API to use integer as key in this PR: https://github.com/odoo/odoo/pull/145765
Then, during the backport OCO made a good point: https://github.com/odoo/odoo/pull/149416#discussion_r1465235277

I complied and forgot about this constraints. -_-"

To me, if you want the minimal diff, you can just change `[current_move_id.id]` to `current_move_id.ids`.
Otherwise, Julien's proposition is right but I would omit the `[]` to make that a generator instead of a list as we don't use it directly but that just nitpick.

Reply:
Updated